### DISCUSSION TITLE
Denotation of Delta

### DISCUSSION THREAD

**Discussion by tomjaguarpaw** [Mikolaj/mostly-harmless#18] [Ideas]
These are my ideas about how to understand the denotation of Delta and its connection to differentiation.

### Delta denotation

Firstly, I think we should separate the denotation of Delta from what it represents in the context of our algorithm, so

* A Delta denotes a mapping from some finite collection of keys each to a Float value. The denotation for `d : Delta` is given by `eval 1 d emptyMap` (or the meta-language version thereof)

[EDIT: added "finite" and the actual denotation]

Nothing to do with derivatives here!  A Delta could perfectly well be used for some other purpose than a differentiation algorithm. In such a context the bullet above would still hold.

### Differentiation algorithm

Now, in our algorithm (i.e. Fig. 9), if we start with an expression `e : Float` we produce an expession `De : (Float, Delta)`.

```
Γ ⊢ e : Float
――――――――――――――――――――――――
DΓ ⊢ De : (Float, Delta)
```

Where `D` is the "curly D, left arrow above" translation from Fig 9 of [our paper](https://www.microsoft.com/en-us/research/uploads/prod/2021/08/higher-order-ad.pdf). `DΓ` means "apply `D` to all the components of the context `Γ`". I am eliding the monad `M` in the types. I hope what I am saying remains clear with `M` left implicit, but I can add it back in if necessary.

### Differentation property

[Updated. See below for original.]

The `De` that our algorithm (i.e. Fig 9) produces has an important property that connects it to differentiation. Consider the result of our algorithm:

```
Γ ⊢ e : Float
――――――――――――――――――――――――
DΓ ⊢ De : (Float, Delta)
```

*The differentiation property*:

Call the following expression `De_plugged`

```
let x1 = (x1, dx1)
    ...
    xn = (xn, dxn)
in snd De
```

(N.B. these variable bindings are non-recursive and shadow) where `{ x1 : Float, ... xn : Float } = Γ`. The denotation property is that

```
Γ, dx1 : Delta, ..., dxn : Delta ⊢ De_plugged : Delta
```

is equal to *in denotation* (i.e. in the denotation defined above) to

```
Γ, dx1 : Delta, ..., dxn : Delta ⊢ Scale (de/dx1) dx1 `Add` ... `Add` Scale (de/dxn) dxn  : Delta
```

if the `Γ ⊢ de/dxi : Float` are expressions that calculate the (partial) derivatives of `e` with respect to each `xi` (at point `Γ`). (For each `i`, `de/dxi` is a metavariable standing for such an expression. Its component parts as a sequence of symbols (`d`, `e`, `/`, `x`, `i`) have no special meaning here.)

This is closely related to Fig. 14 of our paper, but is generalised to expressions rather than just functions. (Fig. 14 applies to functions with range Float^m and the above only deals with expressions of type Float, but the above could easily be generalised to expressions of type Float^m.)

---

### Linearity property

[EDIT: No, we don't need this: it comes for free from the new "full Jacobian" formulation above.]

We probably also need a linearity property that I don't really know how to state formally but informally the property is that

1.  if we scale all `Delta`s in `Γ` by some amount  `x` then the denotation, `Delta` also scales by `x`, and
2. if we add `Delta`s in two contexts of type `Γ` then the resulting `Delta` is equivalent (in denotation) to the sum of the two resulting `Delta` from the two contexts separately.

### Differentation property

The `De` that our algorithm produces has an important property that connects it to differentiation. If `e` has a free variable `x` then our algorithm gives us:

```
Γ, x : Float ⊢ e : Float
――――――――――――――――――――――――――――――――――――――――――――
DΓ, x : (Float, Delta) ⊢ De : (Float, Delta)
```

*The differentiation property*: Suppose we "plug in" `0` (i.e. `Zero`) for all the `Delta`s in `Γ` and we "plug in" `d : Delta` for the one in `x : (Float, Delta)`. Then the `Delta` that "comes out of" the second component of `De` will be (in denotation) `d` with all its values scaled by `de/dx` (evaluated at `Γ, x`).

I'm not sure how to formally state "plug in" and "comes out of" but the former is something to do with substitution (and the latter is probably just applying projection of the second component of a pair).

**Comment by Mikolaj** [comment]
> I think we should separate the denotation of Delta from what it represents in the context of our algorithm,

Which algorithm, though? Because there are two: one for assigning a delta expression to an objective function (turning a function on floats into a function on dual numbers), another for computing the gradient of an objective function based on the delta expression from the dual number representation of the function.

My semantics assumes the first algorithm [edit: the translation from Fig 9] and constitutes a property (a specification) of the second algorithm [edit: the wrapper from Fig 11 and the `eval` function it references] (and of the third that computes the forward derivative of an objective function via its delta expression, see `evalBindingsForward`).

**Reply by tomjaguarpaw** [reply]
When I said "our algorithm" I was talking specifically about the translation from Fig 9 of [our paper](https://www.microsoft.com/en-us/research/uploads/prod/2021/08/higher-order-ad.pdf), so I think that corresponds to your "first algorithm". By "second algorithm" I think you refer to the "wrapper" in Fig 11, is that right?

**Reply by Mikolaj** [reply]
That's correct.

**Reply by Mikolaj** [reply]
I certainly agree with first two of your four sections. The third and fourth are harder to validate, but look fine, too. However, they are not definitions, either formal or informal, but only statements of properties. Do you think they each or together uniquely determine the mapping from keys to floats a delta expression signifies (in the context of `e` and `Gamma`)?

**Reply by tomjaguarpaw** [reply]
> Do you think they each or together uniquely determine the mapping from keys to floats a delta expression signifies

Yes, I think so, because you can use linearity to uniquely determine what happens to the combination in terms of what happens to the individuals. What happens to the individuals is unique because the derivative is unique.

**Reply by tomjaguarpaw** [reply]
(By "combination" I mean "plugging in" arbitrary Deltas, and by "individuals" I mean plugging in Deltas all of which except for one are `Zero`.)

**Reply by Mikolaj** [reply]
Actually, I now think your third section is wrong [edit: I should have said 'misguided' or 'overcomplicated'], because it doesn't mention the `Var` constructor of `Delta` and the differentiation is with respect to the `Var` variables (previously, the one-hots), not `x` variables. In particular, `e` in the wrapper in Fig 11 is of function type and the differentiation is wrt the domain of that function type, not wrt the free variables (I'm not even sure there can be any, but perhaps it's harmless; in my implementation there are none).

**Reply by Mikolaj** [reply]
I think one indeed accesses these differentiable "variables" using normal `x` variables in the surface language. But not any normal `x` variable one can differentiate with respect to. Not even any normal variable of type `Float`, as what you wrote suggests. Only those that correspond to the argument type in the function type of `e`. 

I'm not sure how deep that distinction is, but I feel it may be related to the fact that a delta expression is generated not only from `e` and `Gamma`, but also from the valuation of the variables in the argument type of the function type of `e`. Which is why `Delta` needs to be recomputed from scratch at each gradient descent iteration, even though only the value of parameters (of the argument type) changes, not `e` nor `Gamma`. This boggles my mind, too.

**Reply by tomjaguarpaw** [reply]
> Actually, I now think your third section is wrong

Can you give an example of where my stated "differentiation property" fails to hold?

**Reply by tomjaguarpaw** [reply]
I'm not sure what you mean in your second comment, but perhaps it helps to note that in the differentiation property that I stated `e` is *not* of function type.

> `Delta` needs to be recomputed from scratch at each gradient descent iteration

Yes, agreed

> even though only the value of parameters (of the argument type) changes, not `e` nor `Gamma`

I'm not sure what you mean `Gamma` does change. It contains the values of the parameters. Perhaps additionally this observation might help:

> all its values scaled by `de/dx` (evaluated at `Γ, x`)

**Reply by Mikolaj** [reply]
> `Gamma` does change. It contains the values of the parameters.

Isn't your `Gamma` (I mean `Γ` from you initial exposition) a typing environment, containing a map from variables to types, not to values? And it's used to assign types to delta expressions, not values.

Also look at the wrapper in Fig 11: the values of parameters are passed in a tuple, not in `Gamma`.

**Reply by tomjaguarpaw** [reply]
You said "I now think your third section is wrong" and by "third section" I assumed you meant "differentiation property" but I'm not sure how Fig 11 is relevant to the "differentiation property". Could you elaborate?

> Isn't your Gamma (I mean Γ from you initial exposition) a typing environment, containing a map from variables to types, not to values?

Yes, it's a typing environment, mapping variables to types, but to consider whether the "differentiation property" holds one has to "plug in" values for the Floats in `Γ` in order to compare the `Delta`s in denotation. I don't know how to describe "plugging in" of Float values in `Γ` formally, but operationally it corresponds to "running" `De` on different values of the "parameters".

**Reply by Mikolaj** [reply]
Yes, I meant the "differentiation property". Fig 11 is relevant, because it shows which functions we intend to differentiate and with respect to what variables. You have it confirmed in Theorem 3, which is stated for a closed term.

> > Actually, I now think your third section is wrong
> 
> Can you give an example of where my stated "differentiation property" fails to hold?

To give an example, I'd need a correct semantics of delta expressions that encompasses free variables, right? Because your property states that this semantics changes in some way given some substitutions. However, I don't have such a semantics. Both my semantics in the code and the one implied by the paper (which is similar) are for closed terms, as in Fig 11 and Theorem 3.

Perhaps the semantics from the paper can be extended to free variables, but I suspect your third section is wrong (I should have said 'misguided' or 'overcomplicated') in trying to do that, at least without first tackling the setting from the paper. I may be wrong. :)

**Reply by Mikolaj** [reply]
BTW, I think the tuples of values and `Var` in Fig 11 is exactly how you decided to model the "plugging in" of parameters in the paper. Of course, you are free to reverse that decision. However, if you don't intend to diverge from the paper, then the lack of `Var` in your third section looks to me "wrong" in the sense of "incorrect" even.

**Reply by Mikolaj** [reply]
Or perhaps the "pluggining in" in your extended setting should substitute also `Var` terms somewhere and in this way your third section can be reconciled with the paper and correct at the same time. This is difficult enough for me that I won't attempt such a reconciliation. If you'd like to try a simpler setting, please let me know, under my definition, what is not clear there.

**Reply by tomjaguarpaw** [reply]
Perhaps let's deal with the `Var` thing first. I don't understand how the absence of `Var` can make it wrong. In fact `Var` is an irrelevant *implementation detail* of `Delta`. It's the *denotation* of `Delta` that matters, and that doesn't have `Var` at all! 

Furthermore, the differentiation property has *nothing* to do with `Delta` at all. The same property holds with `Delta` replaced by any vector space. I could have written:

### Differentation property

The `De` that our algorithm produces has an important property that connects it to differentiation. If `e` has a free variable `x` then our algorithm gives us, for any vector space `V`:

```
Γ, x : Float ⊢ e : Float
――――――――――――――――――――――――――――――――――――――――――――
DΓ, x : (Float, V) ⊢ De : (Float, V)
```

*The differentiation property*: Suppose we "plug in" `0` (of the vector space `V`) for all the `V`s in `Γ` and we "plug in" `v : V` for the one in `x : (Float, V)`. Then the `V` that "comes out of" the second component of `De` will be (in denotation) `de/dx * v` (i.e. scalar multiplication of the vector `v`) (evaluated at `Γ, x`).

Now, of course our algorithm as stated uses `Delta` rather than `V` but it is "polymorphic" in the choice of vector space: we could equally well have chosen `V` rather than `Delta`. There are some technicalities around the monad `M` and how we handle bindings, but I don't see that they can invalidate this line of reasoning.

**Reply by tomjaguarpaw** [reply]
> To give an example, I'd need a correct semantics of delta expressions that encompasses free variables, right?

No, I don't think so.  I wrote

* A Delta denotes a mapping from some collection of keys each to a Float value.

and I really mean that as written. Not "a Delta without free variables denotes ...", but *any* Delta. (In fact a Delta without free variables is always denotationally equivalent to zero, isn't it?)

> Because your property states that this semantics changes in some way given some substitutions.

"Plugging in" needs to be expressed somehow in terms of substitution in our language *not* in `Delta`. I'm not referring to any form of substitution in `Delta`.

**Reply by Mikolaj** [reply]
Yes, let's ignore the monad. `Var` is crucial, because it's used to "plug in", not because the monad uses it. In fact, vector spaces in general don't have any mechanisms for plugging things in, so you need something on top of a term algebra of vector spaces.

I thought I'm helping, but I see I'm actually only confusing things. Let me backtrack. If you sketch an argument that your two properties uniquely (up to sensible isomorphisms) determine the semantics of `Delta`, I will be satisfied (constructivism be damned). In particular, please sketch how "plugging in" should be modelled (i think "coming out of" indeed is just a projection). it seems that your property just above is not enough (and the linearity property doesn't help), because a vector space with reversed signs is just as good a model, so this admits at least two models, not one. The "plugging in" is probably a bigger problem, though, and I hope you'd start understanding my objections once you tackle it.

If you'd like me to answer any of the questions above I haven't yet answered, please point them out --- I'm afraid I'm not helping, so I'd rather avoid distracting you.

**Reply by Mikolaj** [reply]
BTW, we have a confusion regarding free variables, too. In all this discussion, whenever I mentioned free variables, I always meant 'x' variables, not 'Var' variables' and I assumed you need 'x' meta-variables in `Delta` expressions to support your definition (basically, you'd have things like `Scale x Zero`, where 'x' is a variable from the 'e' expresions level). I'd be happy to be proven wrong via a complete definition of the semantics.

**Reply by tomjaguarpaw** [reply]
OK, I think I need to come up with a more precise definition of "plugging" in, otherwise we will probably be confused discussing our own, possibly conflicting, ideas of what it is!

**Reply by tomjaguarpaw** [reply]
I updated "Differentiation property" in [my original post](https://github.com/Mikolaj/mostly-harmless/discussions/18#discussion-3906176). Please let me know what you think.

**Reply by Mikolaj** [reply]
Thank you, that's more readable. You are using Haskell as a meta-language to express substitution for free variables in `De`, correct? This is OK, Haskell is as good a meta-language as anything else. 

Could you rename the 'x' in '(x, d)' to something else? E.g., 'v'?

Could you state, as an assumption, something like

    DΓ ⊢ (v, d) : (Float, Delta)

and then informally derive a version of your

    Γ, x : Float, d : Delta ⊢ De_plugged : Delta

from the derivation rules of the respective systems (informally is fine, it's more about which systems we are using, not which rules exactly).

Perhaps a good candidate to derive is

    DΓ ⊢ De_plugged_without_snd : (Float, Delta)

and project only afterwards?

The exposition is indeed improved, because I'm now able to notice that `de/dx : Float` is not well defined --- namely, it depends on the value of the other variables. The derivative of x*y wrt x is y, so it's not a closed float expression or a float constant, but a float expression with free variable or a function from floats to a float.

**Reply by tomjaguarpaw** [reply]
> You are using Haskell as a meta-language to express substitution for free variables in De, correct?

No, there is no meta-language there, only "our language"[1], i.e. the language of Fig. 1. That is, the `let`s and `snd` there are the `let` and `snd` of "our language", not Haskell. (Actually `de/dx` is a meta-variable, but see below.)

> Could you rename the 'x' in '(x, d)' to something else? E.g., 'v'?

I suppose so, but why? For clear connection to the "mathematical notation" `de/dx` it seems valuable to preserve that variable name.

> Could you state, as an assumption, something like `DΓ ⊢ (v, d) : (Float, Delta)`.

I'm not sure what you mean. That's not an assumption. Neither `v` nor `d` is in `DΓ`. I don't know how to handle your subsequent questions, perhaps because I get stuck at your comment above.

> The exposition is indeed improved, because I'm now able to notice that de/dx : Float is not well defined --- namely, it depends on the value of the other variables.

OK, I think I understand what you mean. I will have another attempt.

[1] According to Fig 1. it does indeed seem to be called "our language"

**Reply by tomjaguarpaw** [reply]
OK, I hope I have addressed your concern regarding the non-closedness of `de/dx` by editing [my original comment](https://github.com/Mikolaj/mostly-harmless/discussions/18#discussion-3906176). Please let me know what you think.

**Reply by Mikolaj** [reply]
Right, I mistakenly thought "our languague" is Haskell, but they are close enough, this is fine. The point is, this is the language to which the source language is translated.

Never mind my objection to `let x = (x, d)` --- in Haskell it would loop, but if our language handles that, I can handle that, too.

Let me backpedal again and ask one question at a time. Here

    Γ, x : Float, d : Delta ⊢ De_plugged : Delta

anything on the left side of `⊢` (if that's the relation on "our language" terms defined in the paper) must be a variable. Are `x` and `d` variables? If not, where is `⊢` defined?

**Reply by tomjaguarpaw** [reply]
Yes, `x` and `d` are variables.

(And I'm not sure if this caution is necessary, but to stave off any potential confusion I will reiterate that there is *no* metalanguage here (besides the metavariable `de/dx`). I am not using "our language" as both a language *and* a metalanguage, I am just using it as a language. "Our language" happens to be both the source and target language of the transformation, but in neither case is there any metalanguage.)

**Reply by Mikolaj** [reply]
So far, so good. What does it mean "A is equal in denotation to B"? In particular, is it a quantification over all possible semantics of `Delta`? E.g., all semantics that have this property are "correct"? Is a denotation of a term with free variables a function from values of the variables to the value of the term?

**Reply by tomjaguarpaw** [reply]
> What does it mean "A is equal in denotation to B"?

It means (if A : Delta and B : Delta) that the denotation of A (as described in "[Delta denotation](https://github.com/Mikolaj/mostly-harmless/discussions/18#discussion-3906176) above) is equal to the denotation of B, i.e. the finite map denoted by A is equal to the finite map denoted by B.

**Reply by tomjaguarpaw** [reply]
> E.g., all semantics that have this property are "correct"?

I'm not sure what this means.

> Is a denotation of a term with free variables a function from values of the variables to the value of the term?

Oh, I think I see what you mean. You mean "what is the denotation of a Delta term with free 'our language' variables?", i.e. `Γ ⊢ d : Delta` for non-empty `Γ`. I think there are two possible answers

1. Yes, as you suggest, choose the denotation to be a function with domain "assignments of variables in `Γ` to values" and codomain `Delta`, or alternatively
2. Avoid talking about the denotation of `Γ ⊢ d : Delta` and instead define two non-closed terms `Γ ⊢ d1 : Delta` and `Γ ⊢ d2 : Delta` to be equal in denotation when all substitutions with closed terms
    * `d1[e1/x1]...[en/xn]` (closed)
    * `d2[e1/x1]...[en/xn]` (closed)

    are equal in denotation.

(I was intending to follow standard type theory practice, but I'm not too familiar with that discipline, so I may be mistaken.)

**Reply by Mikolaj** [reply]
I hope I'm not too annoying. :)

A denotation, or a semantics, is a function F from terms (and possible some context, e.g., proving the valuation of free variables) to values, which in this case are finite maps. Where do you take that function from? Are you assuming such a function exists? Are you saying that each such function has a given property? The F is a free variable in your statement. You need to quantify F or define F to make the statement precise.

**Reply by Mikolaj** [reply]
Regarding free variables, thank you, that makes sense. For the sake of clarity, though, it would be good to choose 1. or 2. I'd advise 1.

**Reply by tomjaguarpaw** [reply]
> I hope I'm not too annoying. :)

I think you are asking reasonable questions since I haven't precisely specified everything I'm talking about.

> Where do you take that function from? Are you assuming such a function exists?

I described it in "[Delta denotation](https://github.com/Mikolaj/mostly-harmless/discussions/18#discussion-3906176)" but perhaps I was not explicit enough what it is. In any case, for "Differentiation property" I am assuming it exists, yes. Its existence is required by the notion of "equivalent in denotation", as I think you are pointing out.

> Are you saying that each such function has a given property? The F is a free variable in your statement. You need to quantify F or define F to make the statement precise.

F is the denotation described in "Delta denotation". If you would like me to define it more precisely I can.

**Reply by tomjaguarpaw** [reply]
In fact, to short circuit the discussion, how about "the denotation F[d] of a closed value d : Delta" is

> eval 1 d emptyMap

**Reply by Mikolaj** [reply]
OK, so let me sketch the outline of your whole argument, as I see it: 

1. Denotation F(d) of a delta expression d is "eval.." 
2. Delta expressions are created using the folowing procedure P(e): start with a source language expression e, do this, do that 
3. If a delta expression is created via P as in 2. it has the following two properties expressed in terms of F from 1.     

Did I get it right? Or is it just a starting point for

4. If G has the same property 3. as F, then G is isomorphic to F.

and/or

5. if Q has the same property 3. as P, then Q is isomorphic to P.

In general, 3 is possibly true (though I'm not done with it) and if so, it's good to know, but what was the initial goal and how do we get there? E.g., my initial goal was to informally (but constructively) define F based solely on P [edit: meaning, no reference to `eval`] and in doing so, provide a specification for eval (and, so, also for my implementation of eval).

**Reply by tomjaguarpaw** [reply]
1. Yes
2. Yes, if you change "Delta expressions are created" to "Delta expressions can be created" (there are other ways of creating them of course)
3. Yes

I'm not trying to address 4 or 5. 1-3 deal with the problem, as I understood it, of last week's meeting where we were trying to understand the meaning of "Delta". If you thought I was trying to address 4 or 5 I can see why what I wrote was confusing!

**Reply by Mikolaj** [reply]
So the topic is more "some fundamental properties of the operational denotation of Delta from the paper" rather than "an independent (perhaps denotational) semantics of delta that sheds light on the operational semantics of Delta in the paper". That's absolutely fine. Indeed I expected the latter.

Let me go back to nit-picking the third section: `Scale (de/dx) d : Delta` is not a legal `Delta`-term, because it can contain, say, `x1`. Do you agree?

**Reply by tomjaguarpaw** [reply]
It is an "our language" expression (possibly) containing `x1` (hence not closed) but I don't know why that makes it not a legal `Delta`-term. (To be clear: `x1` is not a `Delta` `Var`, it's an "our language" variable).

**Reply by Mikolaj** [reply]
In the definition of `Delta` I read `Scale R Delta`. Not `Scale e Delta`, where `e` is the language of "our language" expressions.

**Reply by tomjaguarpaw** [reply]
`Delta` is a data type in "our language", so if `e` is an expression of type `R`/`Float` in "our language" and `d` is an expression of type `Delta` in "our language" then `Scale e d` is an expression of type `Delta` in "our language".

I'm not sure what I'm missing here. It seems like there must be a fundamental misalignment in how we understand these things.

**Reply by tomjaguarpaw** [reply]
> my initial goal was to informally (but constructively) define F

Regarding this, I think that any denotation of `Delta`, call it F, should satisfy

1. F maps Delta to a vector space V
2. F is (mathematically) linear, that is F[Zero] = 0, F[Scale x d] = x * F[d] and F[Add d1 d2] = F[d1] + F[d2]
3. F respects substitution, that is F[Let v d1 d2] = F[d2[d1/v]] (where the substitution is substitution in the "Delta language", not "our language")

Does that help clarify things?

**Reply by Mikolaj** [reply]
> `Delta` is a data type in "our language", so if `e` is an expression of type `R`/`Float` in "our language" and `d` is an expression of type `Delta` in "our language" then `Scale e d` is an expression of type `Delta` in "our language".

You are right, although `Scale (de/dx) d` is not a legal term of the `Delta` language (where `Delta` is seen as a grammar, or a set of values, similarly as it's seen in the definition of `eval` in the paper), it is a legal term of "our language" and that's enough to correctly pose what you stated.

> I'm not sure what I'm missing here. It seems like there must be a fundamental misalignment in how we understand these things.

I think I'm just tripping over the fact that the semantics of `Delta` is not defined as a function from the language of delta expressions to the set of finite maps, but that you go one level up and define properties of "our language" terms that, in turn, denote delta expressions. But that may just be my limited mental capacity. Let me continue asking questions.

**Reply by Mikolaj** [reply]
> Does that help clarify things?

Yes, that's helpful. That would be really deep if it was holding for both the denotation of delta expressions as gradient vectors and as forward derivative scalars. Does it? I think it does hold for forward derivatives, because my code for forward derivatives does something like that (though it handles `Let`s by-value, not by-name; I think I even claim in my writeup that the code for gradients can't be by-value, but has to be by-name).

**Reply by tomjaguarpaw** [reply]
I think it *should* hold for both, in fact I would be very surprised if it didn't, but I haven't checked.

**Reply by Mikolaj** [reply]
I can't find a counterexample. That's really spot on. Do you intend to use this formulation in the main article? Would you like my comments at some point?

**Reply by tomjaguarpaw** [reply]
I haven't proposed it to any of the others yet so perhaps it should be brought up for discussion at a forthcoming meeting (although I won't attend tomorrow).

**Reply by tomjaguarpaw** [reply]
I have written this up in LaTeX: https://github.com/neel-krishnaswami/higher-order-ad/blob/e6b1c69964a7421c6b5b36f9ac000d02a8c6eed1/more.tex

**Comment by Mikolaj** [comment]
For comparison, below is the latest version [edit: this one from Apr 2022] of my formulation of the semantics of delta-expressions. It defines the denotation referring solely on the objective function from which the delta expression is derived and not referring to the `eval` function from the paper nor, in any other way, to the syntax of the delta expression in hand.

![image](https://user-images.githubusercontent.com/281893/162410920-c5a3b76d-8266-4e04-83b9-ec6187d1eecd.png)
